### ISSUE TITLE
Pod Lifecycle Issue -> Terminating Pod is immediately removed from Backend Pool

### ISSUE THREAD

**Issue by nboeckmann** [Azure/application-gateway-kubernetes-ingress#641]
**Describe the bug**
In #606 I described an issue with the lifecycle where pods were removed to early from the backend pool. The #608 has improved the behavior but we're not there, yet.

When using HPA in the event of scaling down the replicas a pod is deleted which sets it into the Terminating state. As soon as a Pod is terminating it's taken out of the Endpoints (it's not in the Addresses and also not in the NotReadyAddresses anylonger)

The ingress controller must therefor support a grace period for keeping addresses that are pulled of the list of Addresses (Addresses and NotReadyAddresses) from and endpoint to support the connection draining scenario. (Ideally that time would be configurable or taken from the connection drain timeout annotation)

**To Reproduce**
- Create a Deployment with >=2 Replicas
- Create an Ingress definition for the Deployment
- Terminate one of the pods bound to the deployment
- The backend pool is immediately updated

**Ingress Controller details**

> Name:           illocutionary-ibex-ingress-azure-5bb8575895-pl7nz
> Namespace:      default
> Priority:       0
> Node:           aks-nodepool1-20347689-vmss000002/10.255.248.66
> Start Time:     Wed, 30 Oct 2019 10:36:41 +0000
> Labels:         aadpodidbinding=illocutionary-ibex-ingress-azure
>                 app=ingress-azure
>                 pod-template-hash=5bb8575895
>                 release=illocutionary-ibex
> Annotations:    <none>
> Status:         Running
> IP:             10.255.248.67
> IPs:            <none>
> Controlled By:  ReplicaSet/illocutionary-ibex-ingress-azure-5bb8575895
> Containers:
>   ingress-azure:
>     Container ID:   docker://f5c76f0d4a70ae5e6ab8232429aefba73920ac8ceb3da28092569737c5b333b2
>     Image:          mcr.microsoft.com/azure-application-gateway/kubernetes-ingress:0.10.0
>     Image ID:       docker-pullable://mcr.microsoft.com/azure-application-gateway/kubernetes-ingress@sha256:70e49dac590cf45454c5bc0cda38be913a68d6f35e599130d54a4ef6787ef83e
>     Port:           <none>
>     Host Port:      <none>
>     State:          Running
>       Started:      Wed, 30 Oct 2019 10:36:54 +0000
>     Ready:          True
>     Restart Count:  0
>     Liveness:       http-get http://:8123/health/alive delay=15s timeout=1s period=20s #success=1 #failure=3
>     Readiness:      http-get http://:8123/health/ready delay=5s timeout=1s period=10s #success=1 #failure=3
>     Environment Variables from:
>       illocutionary-ibex-cm-ingress-azure  ConfigMap  Optional: false
>     Environment:
>       AZURE_CONTEXT_LOCATION:  /etc/appgw/azure.json
>       AGIC_POD_NAME:           illocutionary-ibex-ingress-azure-5bb8575895-pl7nz (v1:metadata.name)
>       AGIC_POD_NAMESPACE:      default (v1:metadata.namespace)
>     Mounts:
>       /etc/appgw/azure.json from azure (rw)
>       /var/run/secrets/kubernetes.io/serviceaccount from illocutionary-ibex-sa-ingress-azure-token-2b4pc (ro)
> Conditions:
>   Type              Status
>   Initialized       True
>   Ready             True
>   ContainersReady   True
>   PodScheduled      True
> Volumes:
>   azure:
>     Type:          HostPath (bare host directory volume)
>     Path:          /etc/kubernetes/azure.json
>     HostPathType:  File
>   illocutionary-ibex-sa-ingress-azure-token-2b4pc:
>     Type:        Secret (a volume populated by a Secret)
>     SecretName:  illocutionary-ibex-sa-ingress-azure-token-2b4pc
>     Optional:    false
> QoS Class:       BestEffort
> Node-Selectors:  beta.kubernetes.io/os=linux
> Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
>                  node.kubernetes.io/unreachable:NoExecute for 300s
> Events:
>   Type    Reason     Age   From                                        Message
>   ----    ------     ----  ----                                        -------
>   Normal  Scheduled  74m   default-scheduler                           Successfully assigned default/illocutionary-ibex-ingress-azure-5bb8575895-pl7nz to aks-nodepool1-20347689-vmss000002
>   Normal  Pulling    74m   kubelet, aks-nodepool1-20347689-vmss000002  Pulling image "mcr.microsoft.com/azure-application-gateway/kubernetes-ingress:0.10.0"
>   Normal  Pulled     74m   kubelet, aks-nodepool1-20347689-vmss000002  Successfully pulled image "mcr.microsoft.com/azure-application-gateway/kubernetes-ingress:0.10.0"
>   Normal  Created    74m   kubelet, aks-nodepool1-20347689-vmss000002  Created container ingress-azure
>   Normal  Started    74m   kubelet, aks-nodepool1-20347689-vmss000002  Started container ingress-azure

**Comment by aleksmark** [comment]
I tried to reproduce this and indeed the pod's IP gets removed from the backend list as soon as the pod enters the Terminated state. 

However, I think if timeouts are properly configured application gateway will not terminate ongoing requests. 

In my test case, I enabled connection draining and set the timeout to 300. Also, set the request timeout to 300:

```  
annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/ssl-redirect: "true"
    appgw.ingress.kubernetes.io/request-timeout: "300"
    appgw.ingress.kubernetes.io/cookie-based-affinity: "true"
    appgw.ingress.kubernetes.io/connection-draining: "true"
    appgw.ingress.kubernetes.io/connection-draining-timeout: "300"
```

I have a pod with a PHP script and a sleep of 250 seconds to simulate a long-running request. I hit that pod with a curl from the outside and right after that I deleted the pod with --grace-period=300`.

The pod IP gets removed from the backend list but the request does not terminate. It takes 250 seconds to finish and returns 200 status code. The pod is then still in Terminating state for about ~50 more seconds before it gets completely killed. During those 50 seconds, it does not take any more requests of course. 



If I reduce the request timeout to 200 seconds, the timeout gets hit it and after ~200 seconds I am getting `504 Gateway Timeout`.

If I reduce connection draining timeout to 200 seconds, the timeout gets hit it and after ~200 seconds I am getting `curl: (52) Empty reply from server`

**Comment by hajdukd** [comment]
@aleks

Ive observed one more weird behaviour.

My setup was similar but Im running multiple httpbins pods.

Timeouts are set to 30sec (grace-period, ag request timeout and ag draining).

When the pod is replaced (i kill it and its recreated) Application Gateway ma return 502s. 

To verify that im running multiple parallel calls that each take 20sec to finish.

This may cause distruptions during deployments.

**Comment by hajdukd** [comment]
Update, it seems that AG is completely unpredictable when it comes to managing connections to terminating pods.

Request timeout is not maintained and instead clients get 499s, instead of 502s like documentation mentions. Additionally Im costantly reproducing a situation where terminating pods keep on getting new traffic despite not passing health probes.

**Comment by akshaysngupta** [comment]
@hajdukd Can you make the following changes and tell if anything improves for you ?

1. Change the Pod (Deployment) spec by adding preStop container life cycle hooks, with a delay (sleep) of 90 seconds.
    Example:
    ```yaml
    kind: Deployment
    metadata:
    name: x
    labels:
        app: y
    spec:
    ...
    template:
        ...
        spec:
        containers:
        - name: ctr
            ...
            lifecycle:
            preStop:
                exec:
                command: ["sleep","90"]
    ```

    The `sleep` command assumes the container is based on Linux. For Windows containers the equivalent command is `timeout`
    The addition of the preStop container life cycle hook will:
    1. delay Kubernetes sending SIGTERM to the container by 90 seconds, but put the pod immediately in "Terminating" state
    1. simultaneously this will also immediately remove the pod from Endpoints list
    1. this will cause AGIC to remove the pod from App Gateway's backend pool
    1. pod will continue to run for the next 90 seconds - giving App Gateway 90 seconds to execute "remove from backend pools" command

1. Add connection draining annotation to Ingress to allow for in-flight connections to complete.
   Example:
    ```yaml
    apiVersion: extensions/v1beta1
    kind: Ingress
    metadata:
      name: websocket-ingress
    annotations:
      kubernetes.io/ingress.class: azure/application-gateway
      appgw.ingress.kubernetes.io/connection-draining: "true"
      appgw.ingress.kubernetes.io/connection-draining-timeout: "30"
    ```

    What this achieves - when a pod is pulled from an App Gateway backend it will disappear from the UI, but existing in-flight connections will not be immediately terminated -- they will be given 30 seconds to complete.
    We believe that the addition of the preStop hook and the connection draining annotation will drastically remove the probability for App Gateway to attempt to connect to a terminated pod.

1. Add terminationGracePeriodSeconds to Pod YAML. This must be set to a value that is greater than the preStop hook wait time.
    ```yaml
    kind: Deployment
    metadata:
    name: x
    labels:
        app: y
    spec:
    ...
    template:
        ...
        spec:
        containers:
        - name: ctr
            ...
        terminationGracePeriodSeconds: 101
    ```
To summarize – we need to add a preStop hook waiting for 90 seconds, termination grace period of at least 90 seconds, and connection draining timeout of about 30 seconds.  All of these values should be adopted to the system being tuned of course.

**Comment by hajdukd** [comment]
@akshaysngupta Gonna have a look ASAP. 

Im just curious where does this 90 seconds come from ?

In that approach terminationGracePeriod has to be higher than prehook_timeout + longest_average_response_time (which is system specific). Im aware of what should be the second value, but first one seems highly dependant on something we (users) have no control of ("remove from backend pools" command execution).

**Comment by hajdukd** [comment]
@akshaysngupta I've rerun the test, the number of 502s decreased significantly but did not disappear completely.

**Comment by wapmesquita** [comment]
I am facing the same issue, but my configuration is an internal service with a Deployment and a Service running. When the HPA starts terminating the pods I get a considered amount of 5xx errors.

**Comment by ohads-MSFT** [comment]
When you check the Log Analytics Diagnostic logs for these failing requests, what do you see in the `serverStatus_s` field?
If I understand correctly, it should be blank (or missing), because the server/backend never gets the chance to return anything and the 502 is completely app gateway generated right?
